### PR TITLE
CMake changes to enable use of nmodl as submodule in CoreNEURON

### PR DESCRIPTION
**Pull request by pramodk**
- to enable use of NMODL inside coreneuron as submodule,
        avoid use of PROJECT_SOURCE_DIR directly
   - use of PROJECT_BINARY_DIR consistently; use CMAKE_BINARY_DIR for main components like nmodl binary and libraries only.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -100,43 +113,44 @@ add_subdirectory(ext/pybind11)
...
 
 # =============================================================================
 # Project version from git and project directories
 # =============================================================================
 set(PROJECT_VERSION ${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR})
 
 # generate file with version number from git and nrnunits.lib file path
-configure_file(${CMAKE_CURRENT_SOURCE_DIR}/src/config/config.cpp.in
-               ${CMAKE_CURRENT_BINARY_DIR}/config.cpp @ONLY)
+configure_file(${CMAKE_CURRENT_SOURCE_DIR}/src/config/config.cpp.in ${CMAKE_BINARY_DIR}/config.cpp
```
Comment (ohm314) [inline CMakeLists.txt]: This would place the config.cpp in a new place - top level build dir, is that change intended?

File: src/language/CMakeLists.txt [hunk 2]
```diff
@@ -17,13 +17,13 @@ endif()
 
 add_custom_command(OUTPUT ${AUTO_GENERATED_FILES}
                    COMMAND ${PYTHON_EXECUTABLE}
-                           ARGS ${PROJECT_SOURCE_DIR}/src/language/code_generator.py
+                           ARGS ${NMODL_PROJECT_SOURCE_DIR}/src/language/code_generator.py
                                 ${CODE_GENERATOR_OPTS}
-                           --base-dir ${PROJECT_BINARY_DIR}/src
+                           --base-dir ${CMAKE_BINARY_DIR}/src
```
Comment (ohm314) [inline src/language/CMakeLists.txt]: same as above

File: CMakeLists.txt (hunk 1 above)
Comment (pramodk) [reply CMakeLists.txt]: moved to build/config now

File: docs/Doxyfile.in [hunk 3]
```diff
@@ -814,10 +814,10 @@ WARN_LOGFILE           =
...
 # Note: If this tag is empty the current directory is searched.
 
-INPUT                  = @PROJECT_SOURCE_DIR@/src
-INPUT                 += @PROJECT_SOURCE_DIR@/test
-INPUT                 += @PROJECT_BINARY_DIR@/src/ast
-INPUT                 += @PROJECT_BINARY_DIR@/src/visitors
+INPUT                  = @NMODL_PROJECT_SOURCE_DIR@/src
+INPUT                 += @NMODL_PROJECT_SOURCE_DIR@/test
+INPUT                 += @CMAKE_BINARY_DIR@/src/ast
+INPUT                 += @CMAKE_BINARY_DIR@/src/visitors
```
[the remark this answers is not in the document]
Comment (pramodk) [reply docs/Doxyfile.in]: Initially we were mixing PROJECT_BINARY_DIR and CMAKE_BINARY_DIR.

If you remmeber our conversation last friday, PROJECT_BINARY_DIR is binary directory of last `project` command in CMakeLists.txt. So in case of submodule, there will be new `project` command and hence there will be new directory `<build>/external/nmodl`.

To "simplify", I thought to simply use `CMAKE_BINARY_DIR` everywhere so that all stuff can be found in `<build>` directory itself (as I did for coreneuron + neuron integration).

Reviewing this choice, I now think using `PROJECT_BINARY_DIR` everywhere make more sense to isolate the build scope and not pollute top level project's build directory.

I have done now necessary changes.

[the remark this answers is not in the document]
Comment (ohm314) [reply docs/Doxyfile.in]: Yes that was my thought too. I agree with your new changes
